vlasovsolver/gpu_pitch_angle_diffusion.cpp

```diff
@@ -0,0 +1,743 @@
+   std::vector<Realf> host_sparsity (numberOfLocalCells, 0.0);
+   std::vector<Realf> density_pre_adjust (numberOfLocalCells, 0.0);
+   
+   std::vector<bool> spatialLoopComplete(numberOfLocalCells, false);
+
+   #pragma omp parallel for
+   for (size_t CellIdx = 0; CellIdx < numberOfLocalCells; CellIdx++) { // Iterate over all spatial cells
+
+      const auto CellID                  = LocalCells[CellIdx];
+      SpatialCell& cell                  = *mpiGrid[CellID];
+      
+      host_sparsity[CellIdx]   = 0.01 * cell.getVelocityBlockMinValue(popID);
+
+      // Ensure mass conservation
+      if (getObjectWrapper().particleSpecies[popID].sparse_conserve_mass) {
+         //TODO: parallelize, tho this not usually used (I think) and probably not that expensive
+         Vec vectorSum {0};
+         Vec vectorAdd {0};
+         for (size_t i=0; i<cell.get_number_of_velocity_blocks(popID)*WID3/VECL; ++i) {
+            vectorAdd.load(&cell.get_data(popID)[i*VECL]);
+            vectorSum += vectorAdd;
+            //density_pre_adjust[CellIdx] += cell.get_data(popID)[i];
+         }
+         density_pre_adjust[CellIdx] = horizontal_add(vectorSum);
+      }
+
+      // Diffusion coefficient to use in this cell
+      Real nu0 = 0.0;
+
+      const std::array<Real,3> B = {cell.parameters[CellParams::PERBXVOL] +  cell.parameters[CellParams::BGBXVOL],
+         cell.parameters[CellParams::PERBYVOL] +  cell.parameters[CellParams::BGBYVOL],
+         cell.parameters[CellParams::PERBZVOL] +  cell.parameters[CellParams::BGBZVOL]};
+      const Real Bnorm           = sqrt(B[0]*B[0] + B[1]*B[1] + B[2]*B[2]);
+      const std::array<Real,3> b = {B[0]/Bnorm, B[1]/Bnorm, B[2]/Bnorm};
+      
+      // Save values of b to vector 
+      host_bValues[3*CellIdx] = b[0];
+      host_bValues[3*CellIdx+1] = b[1];
+      host_bValues[3*CellIdx+2] = b[2];
+
+      if (P::PADcoefficient >= 0) {
+         // User-provided single diffusion coefficient
+         nu0 = P::PADcoefficient;
+      } else {
+         // Use nu0 values based on Taniso and betaPara read from file
+         if (!nuArrayRead) {
+            std::cerr<<" ERROR! Attempting to interpolate nu0 value but file has not been read."<<std::endl;
+            abort();
+         }
+
+         // Perform Eigen rotation to find parallel and perpendicular pressure
+         Eigen::Matrix3d rot = Eigen::Quaterniond::FromTwoVectors(Eigen::Vector3d{b[0], b[1], b[2]}, Eigen::Vector3d{0, 0, 1}).normalized().toRotationMatrix();
+         Eigen::Matrix3d Ptensor {
+            {cell.parameters[CellParams::P_11], cell.parameters[CellParams::P_12], cell.parameters[CellParams::P_13]},
+            {cell.parameters[CellParams::P_12], cell.parameters[CellParams::P_22], cell.parameters[CellParams::P_23]},
+            {cell.parameters[CellParams::P_13], cell.parameters[CellParams::P_23], cell.parameters[CellParams::P_33]},
+         };
+         Eigen::Matrix3d transposerot = rot.transpose();
```

Comment:
These are the Eigen commands, so you should be safe to remove the `namespace Eigen` at the top.